code/game/objects/machinery/miner.dm

```diff
@@ -43,21 +52,88 @@
 /obj/machinery/miner/update_icon()
 	switch(miner_status)
 		if(MINER_RUNNING)
-			icon_state = "mining_drill_active"
+			if((mineral_produced == /obj/item/compactorebox/platinum)&&(miner_upgrade_type == MINER_COMPACTOR))
+				icon_state = "mining_drill_active_platinum_[miner_upgrade_type]"
+			else
+				icon_state = "mining_drill_active_[miner_upgrade_type]"
 		if(MINER_SMALL_DAMAGE)
-			icon_state = "mining_drill_braced"
+			icon_state = "mining_drill_braced_[miner_upgrade_type]"
 		if(MINER_MEDIUM_DAMAGE)
-			icon_state = "mining_drill"
+			icon_state = "mining_drill_[miner_upgrade_type]"
 		if(MINER_DESTROYED)
-			icon_state = "mining_drill_error"
+			icon_state = "mining_drill_error_[miner_upgrade_type]"
+
+/obj/machinery/miner/proc/attempt_upgrade(obj/item/plastee,mob/user,params)
+	if(miner_upgrade_type)
+		return FALSE
+	if(user.skills.getRating("construction")<SKILL_CONSTRUCTION_ADVANCED)
+		to_chat(user, "<span class='info'>You can't figure out how to assemble the complex module.</span>")
+		return FALSE
+	var/static/list/modules = list(MINER_RESISTANT = image(icon = 'icons/obj/mining_drill.dmi',icon_state = "mining_drill_reinforceddisplay"),MINER_COMPACTOR = image(icon = 'icons/obj/mining_drill.dmi',icon_state = "mining_drill_compactordisplay"),MINER_OVERCLOCKED = image(icon = 'icons/obj/mining_drill.dmi', icon_state = "mining_drill_overclockeddisplay"))
+	var/choice = show_radial_menu(user, src, modules, require_near = TRUE, tooltips = TRUE)
+	user.visible_message("<span class='notice'>[user] begins attaching a module to [src]'s sockets.</span>")
+	if(!do_after(user, 15 SECONDS, TRUE, src, BUSY_ICON_BUILD))
+		return FALSE
+	if(!plastee.use(MINER_REQUIRED_PLASTEEL_SHEETS))
+		return FALSE
+
+	switch(choice)
+		if(MINER_RESISTANT)
+			max_miner_integrity = 300
+			miner_integrity = 300
+		if(MINER_COMPACTOR)
+			if(mineral_produced == /obj/structure/ore_box/platinum)
+				mineral_produced = /obj/item/compactorebox/platinum
+			else
+				mineral_produced = /obj/item/compactorebox/phoron
+		if(MINER_OVERCLOCKED)
+			required_ticks = 35
+
+	miner_upgrade_type = choice
+	user.visible_message("<span class='notice'>[user] attaches the [choice] to the [src]!</span>")
+	playsound(loc,'sound/items/screwdriver.ogg', 25, TRUE)
+	update_icon()
 
+/obj/machinery/miner/attackby(obj/item/I,mob/user,params)
+	. = ..()
+
+	if(istype(I,/obj/item/stack/sheet/plasteel))
```

Comment:
From a gameplay perspective this 'makes sense' you are trading plasteel for points. 

However I do not like the 'ease' of which this module gets created from a lore perspective. Complex machinery should not be easily fabricated from sheets of metal.

Reply:
i could make them  be Requisitions orderable modules or be engineering vendor items. But then i feel like it would become too tedious which would make everyone not use them.